# Describing the same data with summary statistics

### Python and R Setup

This setup allows you to use *Python* and *R* in the same notebook.

To set up a similar notebook, see quickstart instructions here:

https://github.com/dmil/jupyter-quickstart



In [1]:
%load_ext rpy2.ipython
%load_ext autoreload
%autoreload 2

%matplotlib inline  
from matplotlib import rcParams
rcParams['figure.figsize'] = (16, 100)

import warnings
from rpy2.rinterface import RRuntimeWarning
warnings.filterwarnings("ignore") # Ignore all warnings
# warnings.filterwarnings("ignore", category=RRuntimeWarning) # Show some warnings

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from IPython.display import display, HTML

In [2]:
%%javascript
// Disable auto-scrolling
IPython.OutputArea.prototype._should_scroll = function(lines) {
    return false;
}

<IPython.core.display.Javascript object>

### Import packages in R

In [3]:
%%R

require('tidyverse')


R[write to console]: Loading required package: tidyverse



── Attaching packages ─────────────────────────────────────── tidyverse 1.3.2 ──
✔ ggplot2 3.4.0      ✔ purrr   0.3.5 
✔ tibble  3.1.8      ✔ dplyr   1.0.10
✔ tidyr   1.2.1      ✔ stringr 1.4.1 
✔ readr   2.1.3      ✔ forcats 0.5.2 
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()


### Read data

In [4]:
%%R

# Read data
df <- read_csv('housing_data.csv')

Rows: 189 Columns: 12
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (1): borough
dbl (11): zip, population, pct_hispanic_or_latino, pct_asian, pct_american_i...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


### R syntax - getting a column
in R, the `$` lets you grab a column of a dataframe

in Python this might be something like `df["pct_below_poverty"]

In [5]:
%%R

# Get a column
df$pct_below_poverty

  [1] 19.69 10.68 25.22 25.68 25.20 12.53 15.74 13.10  7.51 27.25 37.16 16.29
 [13] 14.27 28.50  9.13 34.96 34.92 34.92 19.92 34.02 10.81 21.34 13.64 37.77
 [25] 18.65 21.76 27.14 29.75 17.12 32.86 20.80 14.55 19.09 34.41 36.23 17.82
 [37] 37.14 27.36  8.43 18.51 15.60 15.60 16.71 16.71  7.09 19.42 31.16 10.97
 [49] 22.69 25.27 11.76  9.46 14.83 12.35 20.76 12.79  7.96  5.88 13.45 21.97
 [61] 26.48 35.61  6.01  6.01 17.62 23.54  7.40 16.98 17.25 11.89  9.20  9.24
 [73] 19.84  8.20  9.92 15.38 29.09 33.55 11.51 36.13 10.76 27.67  5.30 28.12
 [85] 12.80 21.04 12.24 12.07  5.21 12.73 13.69 11.31 12.32 18.21 38.14 13.26
 [97]  6.77  3.15 10.37 14.53 21.56  5.32 23.14  8.72 15.03 18.67  7.53 14.71
[109]  8.18 12.63 43.12 16.36  9.33 10.18  9.27  7.39 36.77 11.51 10.82  6.63
[121]  6.37 11.56 15.20 16.85 13.37  3.98 27.58 27.58  5.79  9.20 16.08  5.88
[133]  4.36 10.73 14.33  9.68  3.17 25.47  8.03 23.01  8.07 14.11 20.15 10.14
[145]  8.54 13.06 16.41  9.90 11.08 18.57  9.25  2.45 25.60 11.1

### R syntax - the almighty pipe `%>%`

The pipe (`%>%`) takes the output of the previous function and makes it the input to the next

https://towardsdatascience.com/an-introduction-to-the-pipe-in-r-823090760d64


In [6]:
%%R 

df$pct_below_poverty %>% 
    mean()

# Equivalent to...
mean(df$pct_below_poverty)

[1] 15.87868


In [7]:
%%R

df$pct_below_poverty %>% 
    median()

[1] 13.06


In [8]:
%%R 

df$pct_below_poverty %>% 
    var()

[1] 108.451


In [9]:
%%R 

df$pct_below_poverty %>% 
    sd()

[1] 10.41398


In [10]:
%%R 

df %>% 
    group_by(borough) %>%
    summarize(
        mean=mean(pct_below_poverty), 
        median=median(pct_below_poverty), 
        standard_deviation=sd(pct_below_poverty))

# A tibble: 5 × 4
  borough        mean median standard_deviation
  <chr>         <dbl>  <dbl>              <dbl>
1 BRONX          26.7  28.7               11.6 
2 BROOKLYN       18.6  17.2                8.20
3 MANHATTAN      13.8  11.0                9.41
4 QUEENS         12.0  10.7                9.06
5 STATEN ISLAND  12.0   9.24               6.58


**👉 Try It**
Compare the summary statistics to the distributions in your previous assignment. What story do they tell? What stories do they obscure? Why was it important to plot the data in the case of this dataset? What did you gain from plotting the `pct_below_poverty` distribution in various different ways?

> Write your answer here

> What story do they tell?
The summary statistics signify that generally the levels of poverty are higher in the Bronx than in the other boroughs at an average of about 27% and a mean of about 29%. Even though the distributions suggested it, the mean further affirms that after Bronx, Brooklyn zip codes also have a higher average poverty level than other boroughs. 

>What stories do they obscure?
One story obscured by the summary statistics is the existence of outliers. For instance, in thsi case, Queens has the second lowest mean and median. However, from the distribution charts, especially those that are put together borough by borough, we learn that two of the three zip codes with the highest poverty rates are in Queens. 
Furthermore, as seen in the Datasaurus Dozen and the Anscombe's Quartet examples explored in Dipanjan (DJ) Sarkar's Towards Data Science article, many different distributions could result in similar summary statistics, obscuring the real nature of the dataset.

>Why was it important to plot the data in the case of this dataset? 
Plotting helped me get the bigger picture. Summary statistics are one way to get a general sense of the data. But since plotting usually takes each point into account more, then I am better able to see where the data is concentrated and what outliers exist. Furthermore, it helps to have a visual aid to understand the data.

>What did you gain from plotting the `pct_below_poverty` distribution in various different ways?
In some cases, new ways of plotting the data revealed new information. For instance:
>1) the overall chart may not have revealed that the outliers were in Queens and the Bronx but the borough by borough charts did. 
>2) the jitter plot reduced the crowdedness of the basic dot plot chart and helped me compare the zip codes better
>3) the shapes of the density plot, the violin plot and the bee swarm helped to provide visual support to understand the data